internal/pkg/solver/buildkit_loader.go

```diff
@@ -0,0 +1,75 @@
+package solver
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/talos-systems/bldr/internal/pkg/types"
+	"github.com/talos-systems/bldr/internal/pkg/types/v1alpha2"
+)
+
+// BuildkitFrontendLoader loads packages from buildkit client.Reference
+type BuildkitFrontendLoader struct {
+	*log.Logger
+	Context types.Variables
+	Ref     client.Reference
+	Ctx     context.Context
+}
+
+type packageProcess func(baseDir string, contents []byte)
+
+func (bkfl *BuildkitFrontendLoader) walk(path string, process packageProcess) error {
+	entries, err := bkfl.Ref.ReadDir(bkfl.Ctx, client.ReadDirRequest{
+		Path: path,
+	})
+
+	if err != nil {
+		return fmt.Errorf("error readdir %q: %w", path, err)
+	}
+
+	for _, entry := range entries {
+		if os.FileMode(entry.GetMode())&os.ModeDir > 0 {
+			if err = bkfl.walk(filepath.Join(path, entry.GetPath()), process); err != nil {
+				return err
+			}
+		} else if entry.GetPath() == "pkg.yaml" {
```

Comment:
nit, seems like we should have `pkg.yaml` better defined as a constant or something.

Reply:
yep, thanks!